lib/grandpa/tracker_commits.go

```diff
@@ -0,0 +1,117 @@
+	// map of commit block hash to data
+	// data = message + tracking linked list element pointer
+	mapping map[common.Hash]commitMessageMapData
+	// double linked list of block hash
+	// to track the order commit messages were added in.
+	linkedList *list.List
+	capacity   int
+}
+
+type commitMessageMapData struct {
+	message *CommitMessage
+	// element contains a block hash value.
+	element *list.Element
+}
+
+// newCommitsTracker creates a new commit messages tracker
+// with the capacity specified.
+func newCommitsTracker(capacity int) commitsTracker {
+	return commitsTracker{
+		mapping:    make(map[common.Hash]commitMessageMapData, capacity),
+		linkedList: list.New(),
+		capacity:   capacity,
+	}
+}
+
+// add adds a commit message to the commit message tracker.
+// If the commit message tracker capacity is reached,
+// the oldest commit message is removed.
+func (ct *commitsTracker) add(commitMessage *CommitMessage) {
+	blockHash := commitMessage.Vote.Hash
+
+	data, has := ct.mapping[blockHash]
+	if has {
+		// commit already exists so override the commit for the block hash;
+		// do not move the list element in the linked list to avoid
+		// someone re-sending the same commit message and going at the
+		// front of the list, hence erasing other possible valid commit messages
+		// in the tracker.
+		data.message = commitMessage
+		ct.mapping[blockHash] = data
+		return
+	}
+
+	// add new block hash in tracker
+	ct.cleanup()
+	element := ct.linkedList.PushFront(blockHash)
+	data = commitMessageMapData{
+		message: commitMessage,
+		element: element,
+	}
+	ct.mapping[blockHash] = data
+}
+
+// cleanup removes the oldest commit message from the tracker
+// if the number of commit messages is at the tracker capacity.
+// This method is designed to be called automatically from the
+// add method and should not be called elsewhere.
+func (ct *commitsTracker) cleanup() {
+	if ct.linkedList.Len() < ct.capacity {
+		return
+	}
+
+	oldestElement := ct.linkedList.Back()
+	ct.linkedList.Remove(oldestElement)
+
+	oldestBlockHash := oldestElement.Value.(common.Hash)
+	delete(ct.mapping, oldestBlockHash)
+}
+
+// delete deletes all the vote messages for a particular
+// block hash from the vote messages tracker.
+func (ct *commitsTracker) delete(blockHash common.Hash) {
+	data, has := ct.mapping[blockHash]
+	if !has {
+		return
+	}
+
+	ct.linkedList.Remove(data.element)
+	delete(ct.mapping, blockHash)
+}
+
+// getMessageForBlockHash returns the commit messages
+// for a particular block hash from the tracker.
+// It returns nil if the block hash does not exist.
+func (ct *commitsTracker) getMessageForBlockHash(
+	blockHash common.Hash) (message *CommitMessage) {
+	data, ok := ct.mapping[blockHash]
+	if !ok {
+		return nil
+	}
+
+	return data.message
+}
+
+// getMessages returns all the commit messages contained in the
+// tracker in the form of slice of commit messages.
+func (ct *commitsTracker) forEach(f func(message *CommitMessage)) {
```

Comment:
Could we not just return `[]*CommitMessage` in this case rather than requiring us to pass in a callback function?  You should be able to just iterate through the list like:
```
for e := l.Front(); e != nil; e = e.Next() {
	// do something with e.Value
}
```

Reply:
It was originally `getMessages` returning a slice of messages assembled from the mapping.
Depending on the number of tracked messages, this would have some impact on performance.
It also suggested there was some order in messages although there wasn't so that's also an element that pushed me towards the current approach.

I also thought about just returning the mapping directly but that feels wrong (separation of concerns, impl details, weird API, unwanted mutations).

I also wrote a benchmark (in `tracker_commits_test.go`) giving out:

```
Benchmark_ForEachVsSlice/forEach
Benchmark_ForEachVsSlice/forEach-8         	     295	   3818911 ns/op	  101656 B/op	     691 allocs/op
Benchmark_ForEachVsSlice/get_messages_for_iterate
Benchmark_ForEachVsSlice/get_messages_for_iterate-8         	     260	   4023858 ns/op	  918320 B/op	     785 allocs/op
```

Reply:
I'd like to see the results of this benchmark if we revise the mapping.

Reply:
I did the revised mapping, and now the for each is worst. Although that's on the same machine, and the older forEach with the older mapping was twice faster and twice less allocations than the newer mapping + copy all messages.

```
Benchmark_ForEachVsSlice/forEach
Benchmark_ForEachVsSlice/forEach-8         	     152	   7897919 ns/op	  212625 B/op	    1999 allocs/op
Benchmark_ForEachVsSlice/get_messages_for_iterate
Benchmark_ForEachVsSlice/get_messages_for_iterate-8         	     223	   5273454 ns/op	  947635 B/op	    1363 allocs/op
```

Do you think we should revert back (reset the commit) and keep it as it was?

Reply:
I would assume the order needs to be preserved if this is supposed to be a FIFO queue.  I guess the dereferencing of the pointer and typecast is slowing this down.  I would say keep it as is for now since the code is simpler.